src/liballoc/heap.rs

```diff
@@ -163,6 +178,23 @@ mod tests {
     use heap;
 
     #[test]
+    fn allocate_zeroed() {
+        unsafe {
+            let size = 1024;
+            let mut ptr = heap::allocate_zeroed(size, 1);
+            if ptr.is_null() {
+                ::oom()
+            }
+            let end = ptr.offset(size as isize);
+            while ptr < end {
+                assert_eq!(*ptr, 0);
+                ptr = ptr.offset(1);
+            }
+            heap::deallocate(ptr, size, 1);
```

Comment:
This is the buggy line, `ptr` is changed above. This is freeing a pointer to the end, not a pointer to the base.